R/multivariate_karcher_mean.R

```diff
@@ -1,124 +1,165 @@
-  }
-
-  mu = q[, , 1]
-  bmu = beta[, , 1]
-  delta = 0.5
-  tolv = 1e-04
-  told = 5 * 0.001
-  itr = 1
-  sumd = rep(0, maxit + 1)
-  sumd[1] = Inf
-  betan = array(0, c(n, T1, N))
-  qn = array(0, c(n, T1, N))
-  normbar = rep(0, maxit + 1)
-  if(ms == "median"){ #run for median only, saves memory if getting mean
-    d_i = rep(0,N) #include vector for norm calculations
-    v_d = array(0, c(n, T1, N)) #include array to hold v_i / d_i
-  }
+  dims <- dim(beta)
+  L <- dims[1] # Dimension of codomain
+  M <- dims[2] # Size of the evaluation grid
+  N <- dims[3] # Sample size
+  q <- array(dim = c(L, M, N)) # Array for hosting SRVFs
 
+  preproc <- furrr::future_map(1:N, \(n) {
+    beta1 <- beta[ , , n]
+    out <- curve_to_q(beta1, scale = scale)
+    q1 <- out$q
+    if (scale)
+      beta1 <- beta1 / out$len
+    list(q1 = q1, beta1 = beta1)
+  }, .options = furrr::furrr_options(seed = TRUE))
 
-  cat("\nInitializing...\n")
-  gam = matrix(0,T1,N)
-  for (k in 1:N) {
-    out = find_rotation_seed_unqiue(mu, q[, , k], 'O', FALSE, FALSE, lambda)
-    gam[, k] = out$gambest
+  for (n in 1:N) {
+    q[ , , n] <- preproc[[n]]$q1
+    beta[ , , n] <- preproc[[n]]$beta1
   }
 
-  gam = t(gam)
-  gamI = SqrtMeanInverse(t(gam))
-  bmu = group_action_by_gamma_coord(bmu, gamI)
-  mu = curve_to_q(bmu, FALSE)$q
-  mu[is.nan(mu)] <- 0
+  # Initialize mean as the pointwise mean of the curves
+  # AST: improve with medoid, meaning compute distance matrix first.
+  qmean <- rowMeans(q, dims = 2)
+  betamean <- rowMeans(beta, dims = 2)
+  delta <- 0.5
+  tolv <- 1e-04
+  told <- 5 * 0.001
+  itr <- 1
+  sumd <- rep(0, maxit + 1)
+  sumd[1] <- Inf
+  betan <- betat <- array(0, c(L, M, N))
+  qn <- qt <- array(0, c(L, M, N))
+  normbar <- rep(0, maxit + 1)
+  if (ms == "median") {
+    # run for median only, saves memory if getting mean
+    d_i <- rep(0, N) #include vector for norm calculations
+    v_d <- array(0, c(L, M, N)) # include array to hold v_i / d_i
+  }
 
   while (itr < maxit) {
-    cat(sprintf("Iteration: %d\n", itr))
-    mu = mu
-
-    for (i in 1:N) {
-      q1 = q[, , i]
+    cli::cli_alert_info("Iteration {itr}/{maxit}...")
 
-      out = find_rotation_seed_unqiue(mu,q1, 'O', FALSE, FALSE, lambda)
-      dist = sqrt(sum((mu-out$q2best)^2)/T1)
+    alignment_step <- furrr::future_map(1:N, \(n) {
+      beta_i <- beta[ , , n]
+      out <- calc_shape_dist(
+        beta1 = betamean,
+        beta2 = beta_i,
+        mode = "O",
+        rotation = rotation,
+        scale = scale
+      )
+      list(d = out$d, q2n = out$q2n, beta2n = out$beta2n)
+    }, .options = furrr::furrr_options(seed = TRUE))
 
-      qn[, , i] = out$q2best
-      betan[, , i] = q_to_curve(out$q2best)
-
-      if(ms == "median"){ #run for median only, saves computation time if getting mean
-        d_i[i] = sqrt(innerprod_q2(qn[,,i], qn[,,i])) #calculate sqrt of norm of v_i
-      }
-      sumd[itr + 1] = sumd[itr + 1] + dist^2
+    for (n in 1:N) {
```

Comment:
The median computation is wrong, please refer to `curve_karcher_mean` for the code for the metric.

Reply:
Indeed I will fix this by looking at `curve_karcher_mean()`. Well, that is if we keep both functions.